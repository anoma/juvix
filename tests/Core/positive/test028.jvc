-- functional queues

type list {
  nil : list;
  cons : any -> list -> list;
};

def hd := \l case l of { cons x _ := x };
def tl := \l case l of { cons _ x := x };

def rev' := \l \acc case l of {
  nil := acc;
  cons h t := rev' t (cons h acc)
};

def rev := \l rev' l nil;

type Queue {
  queue : list -> list -> Queue;
};

def fst := \q case q of { queue x _ := x };
def snd := \q case q of { queue _ x := x };

def front := \q hd (fst q);

def pop_front := \q
  let q' := queue (tl (fst q)) (snd q) in
  case fst q' of {
    nil := queue (rev (snd q')) nil;
    _ := q'
  };

def push_back := \q \x case fst q of {
  nil := queue (cons x nil) (snd q);
  _ := queue (fst q) (cons x (snd q))
};

def is_empty := \q case fst q of {
  nil := case snd q of { nil := true; _ := false };
  _ := false
};

def empty := queue nil nil;

def g := \q \acc if is_empty q then acc else g (pop_front q) (cons (front q) acc);
def f := \n \q if n = 0 then g q nil else f (n - 1) (push_back q n);

f 100 empty
