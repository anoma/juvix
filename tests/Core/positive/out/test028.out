cons 1 (cons 2 (cons 3 (cons 4 (cons 5 (cons 6 (cons 7 (cons 8 (cons 9 (cons 10 (cons 11 (cons 12 (cons 13 (cons 14 (cons 15 (cons 16 (cons 17 (cons 18 (cons 19 (cons 20 (cons 21 (cons 22 (cons 23 (cons 24 (cons 25 (cons 26 (cons 27 (cons 28 (cons 29 (cons 30 (cons 31 (cons 32 (cons 33 (cons 34 (cons 35 (cons 36 (cons 37 (cons 38 (cons 39 (cons 40 (cons 41 (cons 42 (cons 43 (cons 44 (cons 45 (cons 46 (cons 47 (cons 48 (cons 49 (cons 50 (cons 51 (cons 52 (cons 53 (cons 54 (cons 55 (cons 56 (cons 57 (cons 58 (cons 59 (cons 60 (cons 61 (cons 62 (cons 63 (cons 64 (cons 65 (cons 66 (cons 67 (cons 68 (cons 69 (cons 70 (cons 71 (cons 72 (cons 73 (cons 74 (cons 75 (cons 76 (cons 77 (cons 78 (cons 79 (cons 80 (cons 81 (cons 82 (cons 83 (cons 84 (cons 85 (cons 86 (cons 87 (cons 88 (cons 89 (cons 90 (cons 91 (cons 92 (cons 93 (cons 94 (cons 95 (cons 96 (cons 97 (cons 98 (cons 99 (cons 100 nil)))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))
